ocaml/testsuite/tests/typing-layouts/mixed_constructor_arguments_alpha.ml

```diff
@@ -0,0 +1,391 @@
+                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: Expected all flat constructor arguments after non-value argument,
+       float#, but found boxed argument, float.
+|}];;
+
+type t_ext += A of float * float# * float * int
+
+[%%expect{|
+Line 1, characters 14-47:
+1 | type t_ext += A of float * float# * float * int
+                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: Expected all flat constructor arguments after non-value argument,
+       float#, but found boxed argument, float.
+|}];;
+
+(* String can't appear in the flat suffix *)
+type t_cstr_flat_string_bad1 = A of float# * string
+
+[%%expect{|
+Line 1, characters 31-51:
+1 | type t_cstr_flat_string_bad1 = A of float# * string
+                                   ^^^^^^^^^^^^^^^^^^^^
+Error: Expected all flat constructor arguments after non-value argument,
+       float#, but found boxed argument, string.
+|}];;
+
+type t_ext += A of float# * string
+
+[%%expect{|
+Line 1, characters 14-34:
+1 | type t_ext += A of float# * string
+                  ^^^^^^^^^^^^^^^^^^^^
+Error: Expected all flat constructor arguments after non-value argument,
+       float#, but found boxed argument, string.
+|}];;
+
+(* The string can't appear in the flat suffix. *)
+type t_cstr_flat_string_bad2 = A of float# * float# * string
+
+[%%expect{|
+Line 1, characters 31-60:
+1 | type t_cstr_flat_string_bad2 = A of float# * float# * string
+                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: Expected all flat constructor arguments after non-value argument,
+       float#, but found boxed argument, string.
+|}];;
+
+type t_ext += A of float# * float# * string
+
+[%%expect{|
+Line 1, characters 14-43:
+1 | type t_ext += A of float# * float# * string
+                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: Expected all flat constructor arguments after non-value argument,
+       float#, but found boxed argument, string.
+|}];;
+
+(* The int [c] can appear in the flat suffix. *)
+type t_cstr_flat_int = A of float# * float# * int
+
+[%%expect{|
+type t_cstr_flat_int = A of float# * float# * int
+|}];;
+
+type t_ext += A of float# * float# * int
+
+[%%expect{|
+type t_ext += A of float# * float# * int
+|}];;
+
+type t_cstr_flat_int_multi =
+  | A of float# * float# * int
+  | B of int
+  | C of float# * int
+  | D of float# * int * float#
+  | E of int * float# * int * float#
+
+[%%expect{|
+type t_cstr_flat_int_multi =
+    A of float# * float# * int
+  | B of int
+  | C of float# * int
+  | D of float# * int * float#
+  | E of int * float# * int * float#
+|}];;
+
+type t_ext +=
+  | A of float# * float# * int
+  | B of int
+  | C of float# * int
+  | D of float# * int * float#
+  | E of int * float# * int * float#
+
+[%%expect{|
+type t_ext +=
+    A of float# * float# * int
+  | B of int
+  | C of float# * int
+  | D of float# * int * float#
+  | E of int * float# * int * float#
+|}];;
+
+(* Parameterized types *)
```

Comment:
We should have a test for

```ocaml
type 'a t = C of float# * 'a
```
Currently rejected, which seems correct.  I think you already have a variation of this where we learn the layout from a mutually recursive group.

Reply:
Sounds good, done.